File: Week_04/kotlin/src/_763_PartitionLabels.kt
class _763_ {
    class Solution {
        fun partitionLabels(S: String): List<Int> {
            val lastPos = IntArray(26)
            S.forEachIndexed { i, v -> lastPos[v - 'a'] = i }
            val ret = ArrayList<Int>()
            var l = 0
            var r = 0
            S.forEachIndexed { i, v ->
                r = maxOf(r, lastPos[v - 'a'])
                if (i == r) {
                    ret += r - l + 1
                    l = r + 1
                }
            }
            return ret
        }
    }
}

File: Week_04/js/src/_763_PartitionLabels.js
/**
 * @param {string} S
 * @return {number[]}
 */
var partitionLabels = function(S) {
    const lastPos = new Array(26)
    const n = S.length
    const idxa = 'a'.codePointAt(0)
    for (let i = 0; i < n; i++) {
        lastPos[S.codePointAt(i) - idxa] = i
    }
    const ret = []
    let l = 0, r = 0
    for (let i = 0; i < n; i++) {
        r = Math.max(r, lastPos[S.codePointAt(i) - idxa])
        if (i == r) {
            ret.push(r - l + 1)
            l = r + 1
        }
    }
    return ret
};